[Leeroy Liu] iP

leeroy999 opened the pull request:
# AilurusPro

>Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

AilurusPro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from here.
2. double-click it. 
3. add your tasks. 
4. let it manage your tasks for you 😉

And it is **FREE**!

_Features:_

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

leeroy999 committed:
Level 1. Greet, Echo, Exit

Commit:
Level 2. Add, List

Commit:
Level 3. Mark as Done

Commit:
Level 4. ToDos, Events, Deadlines

Commit:
A-TextUiTesting

Commit:
Level 5. Handle Errors

Commit:
Level 6. Delete

Commit:
Level 6. Delete

Commit:
Add save functionality to hard drive

Commit:
Add LocalDate parsing of deadline and event dates

Commit:
Fix merge conflicts

Commit:
Refactor OOP classes

Commit:
Refactor packages

Commit:
Add JUnit tests for parsing, tasks and storage

Commit:
Add Javadocs to most public and private methods

Commit:
Fix coding standards

Commit:
Add find command to search for tasks

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Fix merge conflicts

qinguorui2001 commented on src/main/java/ailurus/Ailurus.java:
```diff
@@ -0,0 +1,118 @@
+            }
+        }
+    }
+
```
Hi, i think you can put the switch case  to the parser class (parse function) instead. Because that way can divide the functionality of classes, and makes run() shorter.

greyguy21 replied:
Hi, perhaps you can just use one try-catch block to catch the possible errors since they are all the same kind of exception.

greyguy21 replied:
Hi, perhaps you can just make use of one try-catch block to deal with the possible errors thrown since they are all the same type of exception.

greyguy21 replied:
Hi, perhaps you can just make use of one try-catch block to deal with the possible errors thrown since they are all the same type of exception.

leeroy999 replied:
Thank you for your suggestions! The reason why I placed them in the run function, was because the the ui, storage and tasks are all in the main Ailurus class, and it may not make sense to pass them into a parse function.

Yes thank you! Will remove the try catch blocks.

qinguorui2001 commented on src/main/java/ailurus/AilurusException.java:
```diff
@@ -0,0 +1,59 @@
+        EMPTYLIST, EMPTYTODO, EMPTYEVENT, EMPTYDEADLINE,
+        EMPTYBY, EMPTYAT, BY, AT, NUMBER, NAN, LOAD, FIND,
+    }
+
```
Define error enum outside of this class could be better？

qinguorui2001 commented on src/main/java/ailurus/Parser.java:
```diff
@@ -0,0 +1,95 @@
+        String[] dataArr = data.split("[|]");
+        switch (dataArr[0]) {
+            case "T":
+                if (dataArr.length < 3) {
```
perhaps use a variable "lens" to store the dataArr's length to avoid recalculating?

qinguorui2001 commented on src/main/java/ailurus/Storage.java:
```diff
@@ -0,0 +1,93 @@
+        } catch (IOException e) {
+            throw new AilurusException(AilurusException.Error.LOAD);
+        }
+        return list;
```
use (FileNotFoundException e |  IOException e) could be better?

leeroy999 replied:
Tried that and this is the prompt: `Types in multi-catch must be disjoint: 'java.io.FileNotFoundException' is a subclass of 'java.io.IOException.` So I believe I can just catch for IOException. Thank you!

qinguorui2001 commented on src/main/java/ailurus/task/TaskList.java:
```diff
@@ -0,0 +1,107 @@
+    public Task getTask(int i) {
+        return list.get(i);
+    }
+
```
I think it violets tell don't ask principle, may be you can find what should be done to tasks, like compare task times, compare two tasks content, they can be done using a method defined in tasklist class.

qinguorui2001 commented on src/test/java/AilurusTest.java:
```diff
@@ -0,0 +1,61 @@
+        assertEquals(task2.toString(), task2String);
+        assertEquals(task3.toString(), task3String);
+    }
+
```
i like your tests, it could be better if you divide tests methods into different test packages.

qinguorui2001 reviewed (commented):
Generally, well done! just need to be more careful with the coding standards and functionality of each class.

greyguy21 commented on src/main/java/ailurus/AilurusException.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public String getMessage() {
+        switch (this.error) {
+            case EMPTYLIST:
```
For this, I think there should not be indentation for the case clauses.

greyguy21 commented on src/main/java/ailurus/Parser.java:
```diff
@@ -0,0 +1,95 @@
+package ailurus;
+
+import ailurus.task.*;
```
I think it would be better to import the classes in the task package explicitly for better documentation.

greyguy21 replied:
I think it would be better to import the classes in the task package explicitly for better documentation.

greyguy21 commented on src/main/java/ailurus/Ui.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * Read user input
```
It would be better to use singular form for the verbs here.

greyguy21 replied:
It would be better to use singular form for the verbs here.

greyguy21 commented on src/main/java/ailurus/AilurusException.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    @Override
+    public String getMessage() {
+        switch (this.error) {
```
For switch statements, I think there should not be indentation for the case clauses.

greyguy21 reviewed (commented):
Overall, I think except for some minor coding violations, your code is pretty clean and readable. I like the name you gave the chatbot too 😄

Commit:
refactor: change README.md and fix based on comments of PR.

Commit:
Merge branch 'add-gradle-support' into branch-A-Gradle

Commit:
A-Gradle

Commit:
feat: add JavaFX GUI when running Main, and fix a few minor bugs

Commit:
Update README.md

Commit:
Add initial message and add javadocs

Commit:
Fix gradle entry point for generating JAR file

Commit:
Remove unnecessary imports and add specific imports

Unnecessary imports can add to increase load time and lower efficiency, therefore only necessary packages are imported

Let's,
* replace"*" in testing with assertEqual and assertNotEqual for specific imports only

* remove unused FXML imports such as Label.

Commit:
Add assertions

Assertions are used to define assumptions about the program state so that the runtime can verify them.

Let's,
* assert that the data loaded and retrieved from text file is correct format
* assert that response and images are correctly loaded and not null

Commit:
Cherry-pick "Add assertions"

This reverts commit f02c6b7da07e2067ee7ecd5873e55b6260b27219.

(cherry picked from commit d28bf39dcb2dbdb34ded9e0b47771010fe64650e)

Commit:
Revert "Add assertions"

This reverts commit f02c6b7da07e2067ee7ecd5873e55b6260b27219.

Commit:
Revert "Add assertions"

This reverts commit f02c6b7da07e2067ee7ecd5873e55b6260b27219.

(cherry picked from commit d28bf39dcb2dbdb34ded9e0b47771010fe64650e)

Commit:
Refactor code quality by adhereing to the KISS principle

leeroy999 committed:
Merge pull request #1 from leeroy999/branch-A-Assertions

Add assertions

Commit:
Merge branch 'master' of https://github.com/leeroy999/ip

leeroy999 committed:
Merge pull request #2 from leeroy999/branch-A-CodeQuality

Refactor code quality by adhereing to the KISS principle

Commit:
Merge branch 'master' of https://github.com/leeroy999/ip

Commit:
Merge branch 'master' of https://github.com/leeroy999/ip

Commit:
Add command list and allow deleting and marking mulitple tasks as done.

C-MassOps
↳ Mass operations

Provide a way to perform tasks on multiple items e.g., delete some specific items in one go.

Commit:
Add gradle.yml to set up Continuous Integration

Commit:
Fix bye command and dialog box being cut

Commit:
Update checkstyle and fix style issues

Commit:
Fix checkstyle for tests

Commit:
Fix some styling and remove I/O redirection tests

Commit:
Add background and UI fixes

Commit:
Add user guide and representative screenshot. Fix minor case sensitivity of search.

Commit:
Fix bye command and unloading of storage

Commit:
Remove printing into command line

Commit:
Change GUI to be non-resizeable

leeroy999 committed:
Set theme jekyll-theme-cayman

leeroy999 committed:
Set theme jekyll-theme-merlot

leeroy999 committed:
Set theme jekyll-theme-dinky

Commit:
Change docs/README.md userguide

Commit:
Merge remote-tracking branch 'origin/master'

leeroy999 committed:
Set theme jekyll-theme-hacker

leeroy999 committed:
Set theme jekyll-theme-tactile

leeroy999 committed:
Set theme jekyll-theme-slate

Commit:
Minor fix on README.md

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Minor fix on README.md

Commit:
Minor fix on Timer for exiting

Commit:
Minor fix for checkstyle

Commit:
Fix JavaFX empty commands and disable input after "bye" command

Commit:
Fix Javadoc and update parsing divider
